### DISCUSSION TITLE
How to change the wiring on the AudioSourceSdFat class?

### DISCUSSION THREAD

**Discussion by Sparky54** [pschatzmann/arduino-audio-tools#402] [Q&A] [answered]
I'm using your player-sdfat-a2dp example with a MakePython ESP32 and audio board, and have been successful. However, I had to use wires to connect the ESP32 to the board to match the specifications in your example, as the board and ESP32 don't match up to how you describe here: https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-player/player-sdfat-a2dp. I was wondering if there was a way for me to change this to suite my needs. Thanks!

**Comment by pschatzmann** [comment, marked as answer]
Sure, why don't you just remap the pins so that they match with your board?
- see https://github.com/pschatzmann/arduino-audio-tools/wiki/External-DAC
- see https://github.com/pschatzmann/arduino-audiokit/wiki/Problems-with-the-SD-Drive-or-some-conflicting-Buttons for a description how to remap the SPI pins

**Comment by Sparky54** [comment]
So if I have something like this:

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioLibs/AudioSourceSDFAT.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioKitHAL.h"

#define PIN_AUDIO_KIT_SD_CARD_CS 22  
#define PIN_AUDIO_KIT_SD_CARD_MISO 19
#define PIN_AUDIO_KIT_SD_CARD_MOSI 23
#define PIN_AUDIO_KIT_SD_CARD_CLK 18


const char *startFilePath="/";
const char* ext="mp3";
AudioSourceSDFAT source(startFilePath, ext);
A2DPStream out = A2DPStream::instance(); 
MP3DecoderHelix decoder;
AudioPlayer player(source, out, decoder);

AudioKit kit;


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);
  
  auto cfg = out.defaultConfig(TX_MODE);
  cfg.name = "BLE Speaker";
  out.begin(cfg);

  // setup player
  player.setVolume(10);
  player.begin();
}


void loop() {
  player.copy();
}

How would I apply the remapping to the player?

**Reply by pschatzmann** [reply]
No - don't use the AudiKit! Just use the example that shows how you change the pins for SPI if this is needed in your case...
For using the custom pins in the player have a look at the AudioKit examples: e.g. https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sdfat-audiokit/player-sdfat-audiokit.ino

**Comment by Sparky54** [comment]
So I would define the pins and then pass them to AudioSourceSDFAT using SdSpiConfig?

**Reply by pschatzmann** [reply]
The different SD card apis only need the CS pin.

**Reply by Sparky54** [reply]
I've tried passing the remapped CS pin directly to AudioSourceSDFAT. I've also tried passing to AudioSourceSDFAT with SdSpiConfig, plus I've also looked all across the repository for different variations of what I've tried. Every time they fail and end with a sd.begin failure.
 
------------------------------------------------------------------------
Here is the latest iteration of my sketch:
 
#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioLibs/AudioSourceSDFAT.h"
#include "AudioCodecs/CodecMP3Helix.h"
 
#define PIN_AUDIO_KIT_SD_CARD_CS 22
#define PIN_AUDIO_KIT_SD_CARD_MISO 19
#define PIN_AUDIO_KIT_SD_CARD_MOSI 23
#define PIN_AUDIO_KIT_SD_CARD_CLK 18
 
 
const char *startFilePath="/";
const char* ext="mp3";
AudioSourceSDFAT source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
A2DPStream out = A2DPStream::instance();  // A2DP input - A2DPStream is a singleton!
MP3DecoderHelix decoder;
AudioPlayer player(source, out, decoder);
 
 
void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
 
  SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);
  auto cfg = out.defaultConfig(TX_MODE);
  cfg.name = "Brayden’s AirPods Pro";
  out.begin(cfg);
 
  // setup player
  player.setVolume(0.3);
  player.begin();
}
 
 
void loop() {
  player.copy();
}
 
------------------------------------------------------------------------
And here is the log:
 
21:48:01.827 -> [W] AudioA2DP.h : 307 - ==> state: Connected
21:48:01.827 -> [1544376][D][BluetoothA2DPSource.cpp:693] bt_app_av_state_connecting(): [BT_AV] bt_app_av_state_connecting evt 0
21:48:01.827 -> [1544379][I][BluetoothA2DPSource.cpp:699] bt_app_av_state_connecting(): [BT_AV] a2dp connected
21:48:01.827 -> [1544388][I][BluetoothA2DPCommon.cpp:277] set_scan_mode_connectable(): [BT_AV] set_scan_mode_connectable false
21:48:02.623 -> [E] AudioSourceSDFAT.h : 89 - sd.begin failed
21:48:02.623 -> [E] SDDirect.h : 231 - Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
21:48:02.623 ->
21:48:02.623 -> Core  1 register dump:
21:48:02.623 -> PC      : 0x40092dcd  PS      : 0x00060530  A0      : 0x801aeb18  A1      : 0x3ffcc3e0 
21:48:02.623 -> A2      : 0x00000000  A3      : 0xfffffffc  A4      : 0x000000ff  A5      : 0x0000ff00 
21:48:02.623 -> A6      : 0x00ff0000  A7      : 0xff000000  A8      : 0x00000000  A9      : 0x00000004 
21:48:02.623 -> A10     : 0x3ffcc9c4  A11     : 0x80000001  A12     : 0x80098930  A13     : 0x3ffcc2d0 
21:48:02.670 -> A14     : 0x00000003  A15     : 0x00060023  SAR     : 0x0000000a  EXCCAUSE: 0x0000001c 
21:48:02.670 -> EXCVADDR: 0x00000000  LBEG    : 0x40092dcd  LEND    : 0x40092ddd  LCOUNT  : 0xffffffff 
21:48:02.670 ->
21:48:02.670 ->
21:48:02.670 -> Backtrace:0x40092dca:0x3ffcc3e00x401aeb15:0x3ffcc3f0 0x401ab8c5:0x3ffcc700 0x400d7472:0x3ffcc7c0 0x400d7895:0x3ffcc890 0x400d65e7:0x3ffcc8b0 0x400d6749:0x3ffcc8d0 0x400e7d76:0x3ffcc930
21:48:02.670 ->
21:48:02.670 ->
21:48:02.670 ->
21:48:02.670 ->
21:48:02.670 -> ELF file SHA256: 0000000000000000
 
------------------------------------------------------------------------
Here is the decoded stack trace:
 
Decoding stack results
0x401aebf5: _svfprintf_r at /builds/idf/crosstool-NG/.build/HOST-i686-w64-mingw32/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/vfprintf.c line 1528
0x401ab9a5: snprintf at /builds/idf/crosstool-NG/.build/HOST-i686-w64-mingw32/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/snprintf.c line 85
0x400d768e: audio_tools::SDDirect ::listDir(char const*) at C:\Users\Brayd\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLogger.h line 60
0x400d7ab1: audio_tools::AudioSourceSDFAT::selectStream(int) at C:\Users\Brayd\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioLibs/SDDirect.h line 34
0x400d67ff: audio_tools::AudioPlayer::begin(int, bool) at C:\Users\Brayd\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioPlayer.h line 131
0x400d6961: setup() at C:\Users\Brayd\Documents\Arduino\sketch_oct09a/sketch_oct09a.ino line 31
0x400e7e1a: loopTask(void*) at C:\Users\Brayd\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.5\cores\esp32\main.cpp line 42
 
------------------------------------------------------------------------
 
Does anything stick out to you as a simple mistake on my part?

**Comment by pschatzmann** [comment]
I don't see anything that is wrong, but I don't understand why you make your life so complicated. Since you already use the standard SPI pins, why don't you use the standard CS pin as well ?

In addition I suggest that you consider this: https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working

**Reply by Sparky54** [reply]
The kit I'm using has an ESP32 that sits on top of an audio board with an sd card module built in. All the SD module pins are the same as the standard SPI pins you mentioned, except for the CS pin, which is pin 22 instead of pin 5. As mentioned initially, I took some jumper wires and connected the board to the ESP32 in the way defined in the player-sdfat-a2dp example and it worked perfectly. However, issues arose when I tried to put the ESP32 directly onto the pins on the audio board as the pins aren't defined in the same way as the example. Trying to change the CS pin through programming hasn't shown any positive results and always ends in the sd.begin failure.

**Comment by pschatzmann** [comment]
I don't see any reason why this should not work - after all the same approach works perfectly with the AudioKit. As described in the link you should have the SD first working outside of my library! 
- Did manage to have the different ESP32 SD or SDMMC examples to work with your hardware ?
- If you want to use the SDFAT library did you manage the have the sketches (provided by that library) working  ?
 
You should ask the supplier of your hardware or the provider of the different SD libraries for help because this has nothing to do with my audio library: I am pretty sure that your problem has already been solved if you search at the right place...
